<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="headless" default="True" />
  <arg name="control_mode" default="0" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="config_dir" default="$(find tiger)/config"/>
  <arg name="type" default="octo" />
  <arg name="onboards_model" default="v1_default" />
  <arg name="direct_model" default="False" />
  <arg name="direct_model_name" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/empty.world" />
  <arg name="sensor_launch_file" default="$(find tiger)/launch/includes/$(arg onboards_model)/sensors.launch.xml"/>
  <arg name="launch_gazebo" default="True" />
  <arg name="spawn_x" default="0.0"/>
  <arg name="spawn_y" default="0.0"/>
  <arg name="spawn_z" default="0.4"/>
  <arg name="spawn_yaw" default="0.0"/>

  ###########  UAV Config  ###########
  <param name="/uav_info/uav_model" value= "32" /> <!-- TODO temp: dragon control -->
  <param name="/tiger/baselink" value="fc" />

  ###########  Motor Config  ###########
  <!-- TODO temp-->
  <rosparam file="$(find tiger)/config/motor_info/MN4010KV475_Afro_15inch.yaml" command="load"/>

  ###########  Battery Config  ###########
  <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/Battery.yaml" command="load"/>

  ########### Servo Config ###########
  <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/Servo.yaml" command="load"/>

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />

    # World Pos Control Mode: 0
    # World Vel Control Mode: 2
    # Local Vel Control Mode: 3
    # Attitude Control Mode: 4
    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" /> #for sensor plugin
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/EgomotionEstimation.yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/simulation/EgomotionEstimation.yaml" command="load" if="$(eval arg('simulation') * (1 - arg('real_machine')))"/>

    ###########  Gimbal Control  ###########
    <param name="control_plugin_name" value ="control_plugin/dragon_gimbal" />
    <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/PositionGimbalControl.yaml" command="load" />
    <remap from="/gimbals_ctrl" to="/tiger/gimbals_ctrl"/>
    <remap from="/joints_ctrl" to="/tiger/joints_ctrl"/>
    <remap from="/joint_states" to="/tiger/joint_states"/>

    <!-- <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/DifferentialFlatnessPidControlConfig.yaml" command="load" /> -->

    ###########  Teleop  ###########
    <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Sensors  ###########
  <include file="$(arg sensor_launch_file)" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="aerial_robot_model" type="servo_bridge_node" name="servo_bridge" output="screen" ns="tiger"/>

  ###########  Transform Control  ###########
  <node pkg="dragon" type="dragon_transform_control_node" name="tiger" output="screen" respawn="true">
    <rosparam file="$(arg config_dir)/$(arg type)/$(arg onboards_model)/AttitudeLQIControl.yaml" command="load" />
    <remap from="joint_states" to="tiger/joint_states"/>
    <param name="kinematic_verbose" value="false" />
    <param name="control_verbose" value="false" />
    <param name="debug_verbose" value="false" />
    <param name="real_machine" value="$(arg real_machine)" />
  </node>

  ###########  Robot Model  ###########
  <include file="$(find tiger)/launch/robot_model.launch" >
    <arg name="type" value="$(arg type)" />
    <arg name="model" value="$(arg onboards_model)"/>
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') + arg('real_machine') > 0)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
    <arg name="direct_model" value="$(arg direct_model)" />
    <arg name="direct_model_name" value="$(arg direct_model_name)" if="$(arg direct_model)"/>
  </include >

  ########## Simulation in Gazebo #########
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(eval arg('simulation') * (1 - arg('real_machine')))" >
    <arg name="model" value="tiger"/>
    <arg name="launch_gazebo" default="$(arg launch_gazebo)" />
    <arg name="gui" value="false" if="$(arg headless)"/>
    <arg name="headless" value="$(arg headless)"/>
    <arg name="worldtype" value="$(arg worldtype)"/>
    <arg name="spawn_x" value="$(arg spawn_x)"/>
    <arg name="spawn_y" value="$(arg spawn_y)"/>
    <arg name="spawn_z" value="$(arg spawn_z)"/>
    <arg name="spawn_yaw" value="$(arg spawn_yaw)"/>
  </include>

</launch>
